include/aws/common/array_list.h

```diff
@@ -0,0 +1,126 @@
+#ifndef AWS_COMMON_ARRAY_LIST_H
+#define AWS_COMMON_ARRAY_LIST_H
+
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/common.h>
+#include <aws/common/error.h>
+#include <stdint.h>
+#include <string.h>
+
+struct aws_array_list {
+    struct aws_allocator *alloc;
+    size_t current_size;
+    size_t length;
+    size_t item_size;
+    void *data;
+};
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    /**
+    * Initializes an array list with an array of void *. In this mode, the array size will grow
+    * by a factor of 2 upon insertion if space is not available. initial_item_allocation is the number of elements
+    * you want space allocated for. item_size is the size of each element in bytes. Mixing items types is not supported
+    * by this API.
+    */
+    AWS_COMMON_API int aws_array_list_init_dynamic(struct aws_array_list *list,
+        struct aws_allocator *alloc, size_t initial_item_allocation, size_t item_size);
+
+    /**
+    * Initializes an array list with a preallocated array of void *. array_size is the total allocation size of raw_array (not the number of elements),
```

Comment:
We could eliminate the error condition (and division) by using an item count instead of an array size here. This also makes the parameters analogous to the dynamic init case.

Reply:
I like that, it's less cumbersome for the API anyways.